eng/native/configureplatform.cmake

```diff
@@ -314,7 +326,7 @@ else()
     endif()
 endif()
 
-if(NOT CLR_CMAKE_HOST_ARCH_WASM)
+if(NOT CLR_CMAKE_HOST_ARCH_WASM AND NOT CLR_CMAKE_TARGET_ANDROID)
```

Comment:
AFAIK, fPIC is required on Android as well and should work, and is working, is it not?

Reply:
Yes I get this warning during the cmake configure stage:

```
  CMake Warning at /Users/alexander/dev/runtime/eng/native/configureplatform.cmake:339 (message):
    PIE is not supported at link time: PIE (C): clang: warning: argument unused
    during compilation: '-pie' [-Wunused-command-line-argument]



    NO_PIE (C): clang: warning: argument unused during compilation: '-no-pie'
```

CMake has special handling for this as documented in https://cmake.org/cmake/help/v3.14/policy/CMP0083.html:

> Note Android platform has a special handling of PIE so it is not required to use the CheckPIESupported module to ensure flags are passed to the linker.

I'll add a comment for that.

Reply:
Interesting, I have not gotten this warning:

<details>
  <summary>click to expand</summary>

```
root@faa60fe68219:/runtime# ROOTFS_DIR=$(realpath /runtime/.tools/android-rootfs/android-ndk-r21/sysroot) ./src/libraries/Native/build-native.sh -cross -arm64  -release
__DistroRid: android.28-arm64
__RuntimeId: android.28-arm64
Set CrossBuild for arm64 build
Setting up directories for build
Checking prerequisites...
/bin/cmake
Commencing build of "native libraries component" for Linux.arm64.Release in /runtime/artifacts/obj/native/Linux-arm64-Release
  Determining projects to restore...
  Restore completed in 569.5 ms for /runtime/eng/empty.csproj.
Invoking "/runtime/eng/native/gen-buildsys.sh" "/runtime/src/libraries/Native/Unix" "/runtime/src/libraries/Native/Unix" "/runtime/artifacts/obj/native/Linux-arm64-Release" arm64 clang "" "" Release ""  -DCLR_ENG_NATIVE_DIR="/runtime/eng/native" -DCMAKE_STATIC_LIB_LINK=0 -DFEATURE_DISTRO_AGNOSTIC_SSL=0
loading initial cache file /runtime/src/libraries/Native/Unix/tryrun.cmake
-- Check for working C compiler: /runtime/.tools/android-rootfs/android-ndk-r21/toolchains/llvm/prebuilt/linux-x86_64/bin/clang
-- Check for working C compiler: /runtime/.tools/android-rootfs/android-ndk-r21/toolchains/llvm/prebuilt/linux-x86_64/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Performing Test C_SUPPORTS_WUNGUARDED_AVAILABILITY
-- Performing Test C_SUPPORTS_WUNGUARDED_AVAILABILITY - Success
-- Looking for include files sys/socket.h, linux/in.h
-- Looking for include files sys/socket.h, linux/in.h - found
-- Performing Test HAVE_IN_PKTINFO
-- Performing Test HAVE_IN_PKTINFO - Success
-- Performing Test HAVE_IP_MREQN
-- Performing Test HAVE_IP_MREQN - Success
-- Performing Test HAVE_FLOCK64
-- Performing Test HAVE_FLOCK64 - Success
-- Looking for O_CLOEXEC
-- Looking for O_CLOEXEC - found
-- Looking for F_DUPFD_CLOEXEC
-- Looking for F_DUPFD_CLOEXEC - found
-- Looking for getifaddrs
-- Looking for getifaddrs - found
-- Looking for lseek64
-- Looking for lseek64 - found
-- Looking for mmap64
-- Looking for mmap64 - found
-- Looking for ftruncate64
-- Looking for ftruncate64 - found
-- Looking for posix_fadvise64
-- Looking for posix_fadvise64 - found
-- Looking for stat64
-- Looking for stat64 - found
-- Looking for vfork
-- Looking for vfork - found
-- Looking for pipe2
-- Looking for pipe2 - found
-- Looking for getmntinfo
-- Looking for getmntinfo - not found
-- Looking for strcpy_s
-- Looking for strcpy_s - not found
-- Looking for strlcpy
-- Looking for strlcpy - found
-- Looking for posix_fadvise
-- Looking for posix_fadvise - found
-- Looking for ioctl
-- Looking for ioctl - not found
-- Looking for sched_getaffinity
-- Looking for sched_getaffinity - found
-- Looking for sched_setaffinity
-- Looking for sched_setaffinity - found
-- Looking for arc4random_buf
-- Looking for arc4random_buf - found
-- Looking for TIOCGWINSZ
-- Looking for TIOCGWINSZ - found
-- Looking for tcgetattr
-- Looking for tcgetattr - found
-- Looking for tcsetattr
-- Looking for tcsetattr - found
-- Looking for ECHO
-- Looking for ECHO - found
-- Looking for ICANON
-- Looking for ICANON - found
-- Looking for TCSANOW
-- Looking for TCSANOW - found
-- Performing Test HAVE_UTSNAME_DOMAINNAME
-- Performing Test HAVE_UTSNAME_DOMAINNAME - Success
-- Performing Test HAVE_STAT_BIRTHTIME
-- Performing Test HAVE_STAT_BIRTHTIME - Failed
-- Performing Test HAVE_STAT_FLAGS
-- Performing Test HAVE_STAT_FLAGS - Failed
-- Looking for lchflags
-- Looking for lchflags - not found
-- Performing Test HAVE_STAT_TIMESPEC
-- Performing Test HAVE_STAT_TIMESPEC - Failed
-- Performing Test HAVE_STAT_TIM
-- Performing Test HAVE_STAT_TIM - Success
-- Performing Test HAVE_STAT_NSEC
-- Performing Test HAVE_STAT_NSEC - Success
-- Performing Test HAVE_DIRENT_NAME_LEN
-- Performing Test HAVE_DIRENT_NAME_LEN - Failed
-- Performing Test HAVE_STATFS_FSTYPENAME
-- Performing Test HAVE_STATFS_FSTYPENAME - Failed
-- Performing Test HAVE_STATVFS_FSTYPENAME
-- Performing Test HAVE_STATVFS_FSTYPENAME - Failed
-- Looking for statfs
-- Looking for statfs - found
-- Check size of struct statfs
-- Check size of struct statfs - done
-- Performing Test HAVE_GNU_STRERROR_R
-- Performing Test HAVE_GNU_STRERROR_R - Success
-- Performing Test HAVE_READDIR_R
-- Performing Test HAVE_READDIR_R - Failed
-- Performing Test KEVENT_HAS_VOID_UDATA
-- Performing Test KEVENT_HAS_VOID_UDATA - Failed
-- Performing Test HAVE_FDS_BITS
-- Performing Test HAVE_FDS_BITS - Failed
-- Performing Test HAVE_PRIVATE_FDS_BITS
-- Performing Test HAVE_PRIVATE_FDS_BITS - Failed
-- Performing Test HAVE_SENDFILE_4
-- Performing Test HAVE_SENDFILE_4 - Success
-- Performing Test HAVE_SENDFILE_6
-- Performing Test HAVE_SENDFILE_6 - Failed
-- Performing Test HAVE_SENDFILE_7
-- Performing Test HAVE_SENDFILE_7 - Failed
-- Looking for clonefile
-- Looking for clonefile - not found
-- Looking for include file sys/sockio.h
-- Looking for include file sys/sockio.h - not found
-- Looking for include file linux/ethtool.h
-- Looking for include file linux/ethtool.h - found
-- Looking for include file sys/poll.h
-- Looking for include file sys/poll.h - found
-- Looking for epoll_create1
-- Looking for epoll_create1 - found
-- Looking for accept4
-- Looking for accept4 - found
-- Looking for kqueue
-- Looking for kqueue - not found
-- Looking for disconnectx
-- Looking for disconnectx - not found
-- Performing Test HAVE_GETNAMEINFO_SIGNED_FLAGS
-- Performing Test HAVE_GETNAMEINFO_SIGNED_FLAGS - Success
-- Performing Test HAVE_SHM_OPEN_THAT_WORKS_WELL_ENOUGH_WITH_MMAP
-- Performing Test HAVE_SHM_OPEN_THAT_WORKS_WELL_ENOUGH_WITH_MMAP - Failed
-- Performing Test HAVE_CLOCK_MONOTONIC
-- Performing Test HAVE_CLOCK_MONOTONIC - Success
-- Performing Test HAVE_CLOCK_REALTIME
-- Performing Test HAVE_CLOCK_REALTIME - Success
-- Looking for mach_absolute_time
-- Looking for mach_absolute_time - not found
-- Looking for futimes
-- Looking for futimes - found
-- Looking for futimens
-- Looking for futimens - found
-- Looking for utimensat
-- Looking for utimensat - found
-- Performing Test BIND_ADDRLEN_UNSIGNED
-- Performing Test BIND_ADDRLEN_UNSIGNED - Success
-- Performing Test IPV6MR_INTERFACE_UNSIGNED
-- Performing Test IPV6MR_INTERFACE_UNSIGNED - Failed
-- Looking for include file sys/inotify.h
-- Looking for include file sys/inotify.h - found
-- Performing Test INOTIFY_RM_WATCH_WD_UNSIGNED
-- Performing Test INOTIFY_RM_WATCH_WD_UNSIGNED - Success
-- Checking prototype getpriority for PRIORITY_REQUIRES_INT_WHO - False
-- Checking prototype kevent for KEVENT_REQUIRES_INT_PARAMS - False
-- Performing Test HAVE_MKSTEMPS
-- Performing Test HAVE_MKSTEMPS - Success
-- Performing Test HAVE_MKSTEMP
-- Performing Test HAVE_MKSTEMP - Success
-- Performing Test HAVE_NETINET_TCP_VAR_H
-- Performing Test HAVE_NETINET_TCP_VAR_H - Failed
-- Performing Test HAVE_NETINET_UDP_VAR_H
-- Performing Test HAVE_NETINET_UDP_VAR_H - Failed
-- Performing Test HAVE_NETINET_IP_VAR_H
-- Performing Test HAVE_NETINET_IP_VAR_H - Failed
-- Performing Test HAVE_NETINET_ICMP_VAR_H
-- Performing Test HAVE_NETINET_ICMP_VAR_H - Failed
-- Looking for include file sys/cdefs.h
-- Looking for include file sys/cdefs.h - found
-- Performing Test HAVE_TCP_H_TCPSTATE_ENUM
-- Performing Test HAVE_TCP_H_TCPSTATE_ENUM - Success
-- Looking for TCPS_ESTABLISHED
-- Looking for TCPS_ESTABLISHED - not found
-- Check size of struct rt_msghdr
-- Check size of struct rt_msghdr - failed
-- Check size of struct rt_msghdr2
-- Check size of struct rt_msghdr2 - failed
-- Check size of struct if_msghdr2
-- Check size of struct if_msghdr2 - failed
-- Looking for include files sys/types.h, sys/sysctl.h
-- Looking for include files sys/types.h, sys/sysctl.h - not found
-- Looking for include files stdint.h, net/if_media.h
-- Looking for include files stdint.h, net/if_media.h - not found
-- Looking for include file linux/rtnetlink.h
-- Looking for include file linux/rtnetlink.h - found
-- Looking for include file linux/can.h
-- Looking for include file linux/can.h - found
-- Looking for getpeereid
-- Looking for getpeereid - not found
-- Looking for getdomainname
-- Looking for getdomainname - found
-- Looking for uname
-- Looking for uname - found
-- Performing Test HAVE_GETDOMAINNAME_SIZET
-- Performing Test HAVE_GETDOMAINNAME_SIZET - Success
-- Looking for inotify_init
-- Looking for inotify_init - found
-- Looking for inotify_add_watch
-- Looking for inotify_add_watch - found
-- Looking for inotify_rm_watch
-- Looking for inotify_rm_watch - found
-- Looking for include file GSS/GSS.h
-- Looking for include file GSS/GSS.h - not found
-- Looking for GSS_SPNEGO_MECHANISM
-- Looking for GSS_SPNEGO_MECHANISM - not found
-- Looking for GSS_KRB5_CRED_NO_CI_FLAGS_X
-- Looking for GSS_KRB5_CRED_NO_CI_FLAGS_X - not found
-- Looking for include file crt_externs.h
-- Looking for include file crt_externs.h - not found
-- Performing Test HAVE_IN_EXCL_UNLINK
-- Performing Test HAVE_IN_EXCL_UNLINK - Success
-- Performing Test HAVE_TCP_H_TCP_KEEPALIVE
-- Performing Test HAVE_TCP_H_TCP_KEEPALIVE - Failed
-- Performing Test HAVE_BUILTIN_MUL_OVERFLOW
-- Performing Test HAVE_BUILTIN_MUL_OVERFLOW - Failed
-- Looking for pthread_setcancelstate
-- Looking for pthread_setcancelstate - not found
-- Found ZLIB: /runtime/.tools/android-rootfs/android-ndk-r21/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libz.a (found version "1.2.7")
-- Performing Test HAVE_UDAT_STANDALONE_SHORTER_WEEKDAYS
-- Performing Test HAVE_UDAT_STANDALONE_SHORTER_WEEKDAYS - Success
-- Looking for ucol_setMaxVariable
-- Looking for ucol_setMaxVariable - found
-- Found OpenSSL: /runtime/.tools/android-rootfs/android-ndk-r21/sysroot/usr/lib/libcrypto.so (found version "1.1.1d")
-- Looking for EC_GF2m_simple_method
-- Looking for EC_GF2m_simple_method - found
-- Looking for SSL_get0_alpn_selected
-- Looking for SSL_get0_alpn_selected - found
-- Configuring done
-- Generating done
-- Build files have been written to: /runtime/artifacts/obj/native/Linux-arm64-Release
Executing cmake --build "/runtime/artifacts/obj/native/Linux-arm64-Release" --target install -j 2
Scanning dependencies of target System.IO.Compression.Native-Static
Scanning dependencies of target System.IO.Compression.Native
[  1%] Building C object System.IO.Compression.Native/CMakeFiles/System.IO.Compression.Native-Static.dir/runtime/src/libraries/Native/AnyOS/zlib/pal_zlib.c.o
[  1%] Building C object System.IO.Compression.Native/CMakeFiles/System.IO.Compression.Native.dir/runtime/src/libraries/Native/AnyOS/zlib/pal_zlib.c.o
```
</details>

note that we also include NDK provided cmake toolchain: https://github.com/dotnet/runtime/blob/963b1587b48ec2fa09cb5bcd0a53460762a459e4/eng/common/cross/toolchain.cmake#L61-L62

Reply:
it's possible that this is because you're targetting Android API level 28 and we're targetting 21.

Reply:
True. The API level just got revved because of the missing `glob(3)` issue described in the comment.

https://github.com/dotnet/runtime/blob/f79afdbc3145285bb38113913bf9ca185cb8c699/eng/common/cross/build-android-rootfs.sh#L23

Reply:
Btw, I have received fatal error when building CoreCLR inside the device (Android emulator):

```sh
# build command
src/coreclr/build-runtime.sh -skipmanaged -cmakeargs -DFEATURE_EVENT_TRACE=0
```

<img width="420" alt="image" src="https://user-images.githubusercontent.com/3840695/77662544-ac947c80-6f84-11ea-8c93-61f51350e00a.png">

Removing this exclusion for Android fixes this build error. so fPIC is required.

Maybe API level difference is not the problem but certain version of clang in your environment does not support `-no-pie` but the later ones do? I am using clang 9.

Reply:
I'm using the clang that ships as part of the Android NDK.

Not sure why you're seeing this, maybe the cmake logic to add the fPIC doesn't work when building on an actual Android?

Reply:
It works when building with fPIC, i.e. when this condition is enabled, but doesn't work when we exclude it. Maybe we can add a small introspection case for compiler's support of `-no-pie` to make both worlds happy, like we added for GCC a while back here: https://github.com/dotnet/runtime/blob/94d35c35c6110abe445b1562cf10689c25fcbc89/eng/native/configurecompiler.cmake#L285

Reply:
@am11 just so I understand correctly, you're not using the clang from NDK but from somewhere else right?

Reply:
On device, yes. In rootfs, no.
So if we comment out this condition:

```cmake
if(NOT CLR_CMAKE_TARGET_ANDROID) # Android requires PIC and CMake handles this so we don't need the check
```

it works on both device (where clang 9.0.1 is acquired from termux) and rootfs (where it is used from NDK). If it is not working in certain coniguration via NDK (as you experienced the `-no-pie` error), I think the right thing to do is to add introspection:

```cmake
check_compiler_flag(-no-pie COMPILER_SUPPORTS_NO_PIE_FLAG) 
```

and make the condition generic:

```diff
+include(CheckCCompilerFlag)

...

-if(NOT CLR_CMAKE_HOST_ARCH_WASM)
-if(NOT CLR_CMAKE_TARGET_ANDROID) # Android requires PIC and CMake handles this so we don't need the check
+if(NOT CLR_CMAKE_HOST_ARCH_WASM AND COMPILER_SUPPORTS_NO_PIE_FLAG)
```

to make both worlds happy. Since I cannot reproduce the `-no-pie` error, if you could confirm that patch is working, I can put up a PR. :)

Reply:
@am11 I tried it out and it doesn't work because it's a linker flag, not a compiler flag.

I think a much easier solution would be this:

```diff
diff --git a/eng/native/configureplatform.cmake b/eng/native/configureplatform.cmake
index 4e25a3df255..7c1e3acf401 100644
--- a/eng/native/configureplatform.cmake
+++ b/eng/native/configureplatform.cmake
@@ -340,6 +340,6 @@ if(NOT CLR_CMAKE_TARGET_ANDROID) # Android requires PIC and CMake handles this s
         message(WARNING "PIE is not supported at link time: ${PIE_SUPPORT_OUTPUT}.\n"
                   "PIE link options will not be passed to linker.")
     endif()
-    set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 endif(NOT CLR_CMAKE_TARGET_ANDROID)
+    set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 endif(NOT CLR_CMAKE_HOST_ARCH_WASM)
```

So we just skip the check on Android and always set `CMAKE_POSITION_INDEPENDENT_CODE`. I checked and this works for NDK-clang and I think should work for normal clang too. Can you verify?

Reply:
@akoeplinger, I did not realize that there is no facility to check linker flags, at least idiomatically and less wordy. I have issued a feature request: https://gitlab.kitware.com/cmake/cmake/-/issues/20513. Fingers crossed. 🤞

* NDK-clang vs. normal clang

seems like we are dealing with this matrix:

|      clang origin     |                  used by                 | supports <br/>-no-pie |
|:---------------------:|:----------------------------------------:|:----------------:|
| ND with API level 21  | Xamarin                                  |         ✘        |
| NDK with API level 28 | eng/common/cross/build-android-rootfs.sh <br/>(see [usage](https://github.com/dotnet/runtime/issues/33237#issuecomment-601054590)) |         ✓        |
| Termux packages       | physical device and emulators            |         ✓        |

your patch satisfies all three. 👍 

a micro nit: I would modify the comment and indent the if block with 4-spaces as rest of the file. 🙂

Reply:
> a micro nit: I would modify the comment and indent the if block with 4-spaces as rest of the file. 🙂

Yep I did that but then the .diff I pasted to explain my plan looked horrendous so I decided against it 😄

Do you want to file a PR? Otherwise I can do it next week.

Reply:
Sure, I am working on a fix for Clear Linux (stock environment has some default CFlags which conflicts with runtime code), touching cmake files, will include it and ping you on the PR. Thanks!